feat: add `status.lastRunAt` field to WorkflowTemplates. Fixes #1915

qingfeng777 opened the pull request:
<!-- markdownlint-disable MD041 -- this is rendered within existing HTML, so allow starting without an H1 -->

<!--

### Before you open your PR

- Run `make pre-commit -B` to fix codegen and lint problems (build will fail).
- [Signed-off your commits](https://github.com/apps/dco/) (otherwise the DCO check will fail).
- Used [a conventional commit message](https://www.conventionalcommits.org/en/v1.0.0/).

### When you open your PR

- PR title format should also conform to [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).
- "Fixes #" is in both the PR title (for release notes) and this description (to automatically link and close the issue).
- Create the PR as draft.
- Once builds are green, mark your PR "Ready for review".

When changes are requested, please address them and then dismiss the review to get it reviewed again.

-->

<!-- Does this PR fix an issue -->

Fixes #1915 

### Motivation

<!-- TODO: Say why you made your changes. -->
It would be useful to know when the last time a WorkflowTemplate was run

### Modifications

<!-- TODO: Say what changes you made. -->

- add WorkflowTemplateStatus.LastRunAt filed to workflow template filed
- enable workflowTemplate support record last run time

<!-- TODO: Attach screenshots if you changed the UI. -->
![img_v3_02ch_489267f4-e9f9-4204-967a-ed83b98d1c2g](https://github.com/argoproj/argo-workflows/assets/16607730/fcbcb369-86ca-4309-afe2-f0909f52fc59)


### Verification

<!-- TODO: Say how you tested your changes. -->
run locally, submit workflow from template and check

<!--
### Beyond this PR

Thank you for submitting this! Have you ever thought of becoming a Reviewer or Approver on the project?

Argo Workflows is seeking more community involvement and ultimately more [Reviewers and Approvers](https://github.com/argoproj/argoproj/blob/main/community/membership.md) to help keep it viable.
See [Sustainability Effort](https://github.com/argoproj/argo-workflows/blob/main/community/sustainability_effort.md) for more information.

-->

Joibel commented:
The implementation of the feature doesn't do what I expected.

Currently, it is only recording uses of WorkflowTemplates when invoked through the user interface or CLI, and will not record uses in many other circumstances:
* Uses from CronWorkflows
* Uses when a workflow is injected via the kubernetes API, or other system which uses this mechanism such as argo-events
* Uses of a template when it is not used through `workflowTemplateRef`.

I am happy, if we document it as so, that this only records `workflowTemplateRef` uses of a WorkflowTemplate. It must  record all uses via `workflowTemplateRef` though. Take a look at [`fetchWorkflowSpec`](https://github.com/argoproj/argo-workflows/blob/1a81dfb305a8576bea276a8cb2d0ad30c03c2cfb/workflow/controller/operator.go#L3821).

I was also expecting this new value to be visible in the user interface, although I might be persuaded that that should be in a separate PR if you have a good reason.

Joibel reviewed (changes requested):
This should be performed by the controller to catch all uses, not by the server.

qingfeng777 commented:
> The implementation of the feature doesn't do what I expected.
> 
> Currently, it is only recording uses of WorkflowTemplates when invoked through the user interface or CLI, and will not record uses in many other circumstances:
> 
> * Uses from CronWorkflows
> * Uses when a workflow is injected via the kubernetes API, or other system which uses this mechanism such as argo-events
> * Uses of a template when it is not used through `workflowTemplateRef`.
> 
> I am happy, if we document it as so, that this only records `workflowTemplateRef` uses of a WorkflowTemplate. It must record all uses via `workflowTemplateRef` though. Take a look at [`fetchWorkflowSpec`](https://github.com/argoproj/argo-workflows/blob/1a81dfb305a8576bea276a8cb2d0ad30c03c2cfb/workflow/controller/operator.go#L3821).
> 
> I was also expecting this new value to be visible in the user interface, although I might be persuaded that that should be in a separate PR if you have a good reason.

Thank you for being so helpful. You were absolutely right. I will make a separate PR to add the Status field and then see how to perform this task with the controller to catch all uses.

Joibel reviewed (changes requested):
By itself, adding the field but not setting it isn't useful, and is possibly misleading as someone might reasonably expect it to get set.

Could you add the code to set this value from the controller to this same PR please.

qingfeng777 commented:
> By itself, adding the field but not setting it isn't useful, and is possibly misleading as someone might reasonably expect it to get set.
> 
> Could you add the code to set this value from the controller to this same PR please.

Thank you for your advice, I will add the code to set this value from the controller to the same PR later.

qingfeng777 commented:
> By itself, adding the field but not setting it isn't useful, and is possibly misleading as someone might reasonably expect it to get set.
> 
> Could you add the code to set this value from the controller to this same PR please.

Hi,

In my first attempt, I updated the Getter in context.go to an Operator (supporting both get and update operations), using the clientSet as a parameter. However, the requeue and get workflow did not function as expected.

In my second attempt, I used NewContextFromClientSet instead of NewContext, to hold both the informer getter and clientSet. Unfortunately, I was unable to retrieve clientSet in some callers of the ValidateWorkflow function.

Ultimately, I implemented the feature like this. I am unsure if there is a better approach. Could you please review it again?

Thank you.

Joibel commented on workflow/templateresolution/context.go:
```diff
@@ -135,6 +162,12 @@ func (ctx *Context) GetTemplateFromRef(tmplRef *wfv1.TemplateRef) (*wfv1.Templat
 		}
 		return nil, err
 	}
+	if !tmplRef.ClusterScope && wftmpl != nil {
```
Here we don't update ClusterWorkflowTemplates (`!tmplRef.ClusterScope`). They should be updated too. It means cwftmplClient which you've added doesn't do any work ever.

qingfeng777 replied:
How about adding the feature of updating ClusterWorkflowTemplates' status in a separate PR?
It will be easier for me to manage

Joibel replied:
I'd want the two merged at the same time, so they'd need to be stacked PRs which is hard to manage too. Get it right for non-cluster and the clusterwftemplates should be pretty simple to add at the end.

qingfeng777 replied:
OK, I got it.

Joibel commented on workflow/templateresolution/context.go:
```diff
@@ -166,6 +199,17 @@ func (ctx *Context) GetTemplateScope() string {
 }
 
+// UpdateTemplateStatus update the WorkflowTemplate.Status of a given WorkflowTemplate.
+func (ctx *Context) UpdateTemplateStatus(wftmple *wfv1.WorkflowTemplate) (*wfv1.WorkflowTemplate, error) {
```
UpdateTemplateStatus should not be capitalised, as it is not necessary to [export](https://go.dev/tour/basics/3) it. Rename to `updateTemplateStatus`.

Joibel commented on workflow/templateresolution/context.go:
```diff
@@ -166,6 +199,17 @@ func (ctx *Context) GetTemplateScope() string {
+	}
+
+	wftmple.Status.LastRunAt = metav1.Now()
+	return ctx.wftmplClient.Update(context.TODO(), wftmple, metav1.UpdateOptions{})
```
This is going to be very hard on the kubernetes API, updating for every time we look at the template, and doing it during execution. I think this needs to go into a deduplicating queue, which is doing rate limited writes to the actual API.
One run of the [dag diamond](https://github.com/argoproj/argo-workflows/blob/main/examples/workflow-template/dag.yaml) called this 33 times.

qingfeng777 replied:
How about setting a ticker to run every ten minutes? The ticker can maintain a map that stores the updated workflows and their corresponding update times, ensuring updates are performed once every ten minutes.

qingfeng777 replied:
Thank you, I will try it later.

Joibel commented on workflow/templateresolution/context.go:
```diff
@@ -166,6 +199,17 @@ func (ctx *Context) GetTemplateScope() string {
+// UpdateTemplateStatus update the WorkflowTemplate.Status of a given WorkflowTemplate.
+func (ctx *Context) UpdateTemplateStatus(wftmple *wfv1.WorkflowTemplate) (*wfv1.WorkflowTemplate, error) {
+	if ctx.wftmplClient == nil {
+		ctx.log.Warnln("while try update template status, wftmplClient not set")
```
This error message is coming out when running the [dag diamond](https://github.com/argoproj/argo-workflows/blob/main/examples/workflow-template/dag.yaml), which probably shouldn't happen. Could you investigate.

Joibel replied:
I believe you run inside a devcontainer. So I assume you're testing with `make start`.

Assuming you are in the argo-workflows checkout directory - the logs will be in `logs/controller.log`. You can run `kubectl create -f examples/dag_diamond.yaml` and I got this error.

qingfeng777 replied:
Thank you, I ran in my machine directly, and got the logs in `logs/controller.log` too.

qingfeng777 replied:
But, I can't reproduce it. Those are what I did.
`
git checkout <target-branch>
make clean
make start UI=true

 kubectl create -f examples/dag-diamond.yaml
    + kubectl create -f examples/dag-diamond.yaml
    workflow.argoproj.io/dag-diamond-67fvs created
    
➜  logs git:(support-record-preset) cat controller.log|grep while
➜  logs git:(support-record-preset)
`
And I checked other files in the `logs`, not found  the error log too.
Did I miss something? or could you recheck it?

qingfeng777 replied:
I tried to reproduce by running dag workflow with a tmpl-ref, with the following commands:
```
kubectl create -f examples/workflow-template/templates.yaml
kubectl create -f examples/workflow-template/dag.yaml
```
but I still couldn't find the error log, so I mark this conversation resolved. If I missed something, feel free to reopen it.
Thank you a lot.

Joibel reviewed (changes requested):
This is a hard problem to solve correctly. The main objection here is how hard this is on the kubernetes API, and how it will slow down execution.

Joibel commented:
I am away for 2 weeks, I will take a look when I can.

qingfeng777 commented:
> I am away for 2 weeks, I will take a look when I can.

Happy holidays!

qingfeng777 commented:
Hi，@Joibel, could you take a look at this PR? thank you.

qingfeng777 committed:
feat: workflow template support record last run time with workqueue. Fixes #1915

Signed-off-by: qingfeng777 <251098199@qq.com>

terrytangyuan reviewed (commented):
I wonder instead of just showing last run, we can keep a history of runs.
